validator_client/src/duties_service/sync.rs

```diff
@@ -0,0 +1,418 @@
+                .await
+        })
+        .await;
+
+    let duties = match duties_response {
+        Ok(res) => res.data,
+        Err(e) => {
+            warn!(
+                log,
+                "Failed to download sync committee duties";
+                "sync_committee_period" => sync_committee_period,
+                "error" => %e,
+            );
+            return Ok(());
+        }
+    };
+
+    info!(log, "Fetched duties from BN"; "count" => duties.len());
+
+    // Add duties to map.
+    let committee_duties = duties_service
+        .sync_duties
+        .get_or_create_committee_duties(sync_committee_period, local_indices);
+
+    // Track updated validator indices & pubkeys.
+    let mut updated_validators = vec![];
+
+    {
+        let mut validator_writer = committee_duties.validators.write();
+        for duty in duties {
+            let validator_duties = validator_writer
+                .get_mut(&duty.validator_index)
+                .ok_or(Error::SyncDutiesNotFound(duty.validator_index))?;
+
+            let updated = validator_duties.as_ref().map_or(true, |existing_duties| {
+                let updated_due_to_reorg = existing_duties.duty.validator_sync_committee_indices
+                    != duty.validator_sync_committee_indices;
+                if updated_due_to_reorg {
+                    warn!(
+                        log,
+                        "Sync committee duties changed";
+                        "message" => "this could be due to a really long re-org, or a bug"
+                    );
+                }
+                updated_due_to_reorg
+            });
+
+            if updated {
+                info!(
+                    log,
+                    "Validator in sync committee";
+                    "validator_index" => duty.validator_index,
+                    "sync_committee_period" => sync_committee_period,
+                );
+
+                updated_validators.push(duty.clone());
+                *validator_duties = Some(ValidatorDuties::new(duty));
+            }
+        }
+    }
+
+    // Spawn background task to fill in aggregator selection proofs.
+    let sub_duties_service = duties_service.clone();
+    duties_service.context.executor.spawn_blocking(
+        move || {
+            fill_in_aggregation_proofs(
+                sub_duties_service,
+                &updated_validators,
+                sync_committee_period,
+            )
+        },
+        "duties_service_sync_selection_proofs",
+    );
+
+    Ok(())
+}
+
+pub fn fill_in_aggregation_proofs<T: SlotClock + 'static, E: EthSpec>(
+    duties_service: Arc<DutiesService<T, E>>,
+    validators: &[SyncDuty],
+    sync_committee_period: u64,
+) {
+    let spec = &duties_service.spec;
+    let log = duties_service.context.log();
+
+    // Generate selection proofs for each validator at each slot, one epoch at a time
+    let start_epoch = spec.epochs_per_sync_committee_period * sync_committee_period;
```

Comment:
Note to self: this should start from the current epoch so we don't waste time computing duties past epochs when we restart

Reply:
Fixed!